### ISSUE TITLE
Document how to run tests in parallel

### ISSUE THREAD

**Issue by mlvandijk** [cucumber/docs#298]
We are getting some questions in Slack on how to run Cucumber-jvm in parallel. This is possible as of v4(?). We should document how to do this correctly. 

Some info: https://cucumber.io/blog/2018/09/24/announcing-cucumber-jvm-4-0-0

http://grasshopper.tech/283/

Note: As I have not had use for running in parallel, I am not familiar with this. Help very much wanted!

**Comment by mlvandijk** [comment]
Denis Sln   [Feb 19th at 11:55]
Hi people!
I'm trying to find best parallel execution solution using cucumber4-java-testng. Currently, I was able to execute scenarios in parallel threads, but I want to try running each scenario in separate JVM to compare performance, however with out-of-box solution I am unable to fork new JVM instances. Tried both surefire and failsafe, no luck? Is it possible to execute scenarios in parallel JVM-s using just a single Cucumber runner and no other 3rd party plug-ins (edited)

16 replies

gabel   [8 days ago]
When you tried surefire and failsafe, did you use `threadCount` or `forkCount`? `forkCount` is the parameter that actually runs a test in a separate JVM. https://maven.apache.org/surefire/maven-failsafe-plugin/examples/fork-options-and-parallel-execution.html

Denis Sln   [8 days ago]
@gabel I tried both, also used combinations. Went through that maven references several times this weekend)

gabel   [8 days ago]
That's odd. I'm in the cucumber 2 world still, so I use an extra maven plugin which creates the runner files properly and reassembles all the reports. https://github.com/temyers/cucumber-jvm-parallel-plugin
temyers/cucumber-jvm-parallel-plugin
Maven plugin to help running Cucumber features in parallel
Stars
122
Language
Java
temyers/cucumber-jvm-parallel-pluginJan 30th, 2015Added by GitHub

gabel   [8 days ago]
But I think that same support is now built into cucumber 4, so I'm not really sure. what the issue is then.

Denis Sln   [8 days ago]
That's what I heard as well and trying to figure out how it works. I switched from Cucumber 1.2.x version. Used it together with Cucables plug-in seems to be doing similar job to the one you are referencing. Generating separate runners

Denis Sln   [8 days ago]
yeah me neither and no warnings or errors

Denis Sln   [8 days ago]
it's just when I'm adding
<systemPropertyVariables>            <databaseSchema>MY_TEST_SCHEMA_${surefire.forkNumber}</databaseSchema>
</systemPropertyVariables> <workingDirectory>FORK_DIRECTORY_${surefire.forkNumber}</workingDirectory>

to POM execution fails

Denis Sln   [8 days ago]
Not sure if that part is a must have for configuration

Denis Sln   [8 days ago]
@mpkorstanje can you please advise?

mpkorstanje   [8 days ago]
Parallelization happens at JUnit method level.

mpkorstanje   [8 days ago]
Or equivalent.

mpkorstanje   [8 days ago]
I do not know if surefire runs those in parallel

Denis Sln   [8 days ago]
I'm feeding scenarios as a separate testNG test (at least to my mind)
@DataProvider(parallel = true)
   public Object[][] scenarios() {
       return super.scenarios();
   }
My expectation is that they run in separate JVM, but OK.. thanks will dig further (edited)

mpkorstanje   [8 days ago]
No. Afaik @DataProvider only ever works in a single  JVM. (edited)

Denis Sln   [7 days ago]
Ok, this is something I wasn't aware about, my failures make sense now. In case anyone got something to suggest/try please go ahead

Denis Sln   [7 days ago]
@grasshopper just came across your article http://grasshopper.tech/283/ seems only threaded parallelism is supported by cucumber out of box.
can you recommend any solution to parallel test in separate JVMs for Cucumber4? Thnx (edited)

**Comment by tsundberg** [comment]
There are at least three different ways to run in parallel
* from a command line by setting the number of threads to use for the execution
  - there are various plugins available for simplifying the usage of the command, at least for Gradle
* through JUnit by using JUnits built in support
  - Maven - by setting maven-surefire-plugin to use a number of threads (https://github.com/cucumber/cucumber-jvm/tree/v4.0.0/junit)
  - Gradle - where the only support is on JUnit class level (resulting in no parallelism since you only have one JUnit class to drive the execution) 
* through TestNG - https://github.com/cucumber/cucumber-jvm/tree/v4.0.0/testng

It would be very nice with executable examples for each, or at least a few, possibilities available in the examples section, https://github.com/cucumber/cucumber-jvm/tree/master/examples

**Comment by mlvandijk** [comment]
Possibly helpful: https://stackoverflow.com/questions/54660808/cucumber-testng-4-0-parallel-execution-defaults-to-10-always

**Comment by mlvandijk** [comment]
Some info for Ruby from Gitter (@enkessler): The parallel_tests gem can run tests in parallel by both feature or by scenario (but not by individual outline row).
https://github.com/grosser/parallel_tests#setup-for-non-rails

**Comment by grasshopper7** [comment]
Had spent some time exploring inbuilt parallel running options in cucumber jvm 4 to migrate an existing project from cucumber-jvm-parallel-plugin. Collated the findings for testng - http://grasshopper.tech/466/ and juint - http://grasshopper.tech/464/. Maybe these help.

**Comment by enkessler** [comment]
As much as it would be nice to have parallel built in, in almost every situation where I need to run in parallel I am simply building run lists and then passing them to a new Cucumber process.

e.g.

files_to_run.txt
```
features/foo.feature:4
features/foo.feature:9
features/foo.feature:13
features/bar.feature:4
features/bar.feature:23
```

`cucumber @files_to_run.txt`

I know that you could probably do the same thing in a script that kicks off runner classes or something, but the command line tends to be more portable across languages.

**Comment by mlvandijk** [comment]
@enkessler In cucumber-jvm we have parallel built in as of v4. We just need to document how to do so.

**Comment by mlvandijk** [comment]
From Slack:

rubytester [23:18]
I would like to run Cucumber tests in parallel ( using `Cucumber 4` `Gradle 4.8` `JUnit 5`). What would be better approach? Is any plugin for that. I’m not using class runner since JUnit 5 doesn’t support annotation `@RunWith` anymore.

mpkorstanje   [6 hours ago]
Use the JUnit Vintage module of JUnit 5 and continue with `@RunWith`.

**Comment by grasshopper7** [comment]
@mlvandijk Is someone working on this?

**Comment by mlvandijk** [comment]
Afaik there is not, so if you want to, please assign yourself and go ahead. Would be much appreciated!

**Comment by grasshopper7** [comment]
@mlvandijk Will work on it. Though I do not have access to assign it to myself.

**Comment by mlvandijk** [comment]
Fixed in #347 thanks to @grasshopper7
